### PR TITLE
Add DT_HALF support for SpaceToDepth on GPU

### LINKED ISSUE #14871: SpaceToDepth for DT_HALF is not supported on GPU
space_to_depth can be placed on GPU for float32, but there is no a GPU kernel for float16

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #14871 where there were no DT_HALF support for SpaceToDepth on GPU.

This fix adds DT_HALF support on GPU and adds aditional test cases.

This fix fixes #14871.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by ygoncharov** [comment]
Does it make sense to update DepthToSpace as well in the same PR?

**Comment by yongtang** [comment]
@ygoncharov Thanks. The PR has been updated with changes in DepthToSpace as well.

**Comment by drpngx** [comment]
@josh11b could you take a look?

File: tensorflow/core/kernels/depthtospace_op.cc [hunk 1]
```diff
@@ -24,6 +24,7 @@ limitations under the License.
 #include "tensorflow/core/kernels/depthtospace_op.h"
 
 #include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
```
Comment (josh11b) [inline tensorflow/core/kernels/depthtospace_op.cc]: Small nit: Prefer to avoid making unrelated formatting changes.

**Review by josh11b — APPROVED** [review]
I think this is fine with just one minor nit.

File: tensorflow/core/kernels/depthtospace_op.cc (hunk 1 above)
Comment (yongtang) [reply tensorflow/core/kernels/depthtospace_op.cc]: Thanks @josh11b. The unrelated formatting changes in those two files have been removed.

**Comment by yongtang** [comment]
Thanks @josh11b for the review. The PR has been updated with suggested changes.
